// @sanity/runtime-cli/dist/cores/blueprints/promote.js
import { confirm } from '@inquirer/prompts';
import { patchConfigFile } from '../../actions/blueprints/config.js';
import { promoteStack } from '../../actions/blueprints/stacks.js';
import { niceId } from '../../utils/display/presenters.js';
export async function blueprintPromoteCore(options) {
    const { log, stackId, auth, flags, deployedStack, blueprint } = options;
    if (!flags['i-know-what-im-doing']) {
        return { success: false, error: 'Seems you do not know what you are doing.' };
    }
    let message = `"${deployedStack.name}" ${niceId(deployedStack.id)}`;
    if (deployedStack.scopeType === 'organization') {
        message = `Stack ${message} is already org-scoped. Promote again?`;
    }
    else {
        message = `Are you sure you want to promote Stack ${message}?`;
    }
    if (!flags.force) {
        const confirmed = await confirm({ message });
        if (!confirmed) {
            return { success: false, error: 'Promotion cancelled by user' };
        }
    }
    try {
        const { ok, error, stack } = await promoteStack({ stackId, auth, logger: log });
        if (!ok) {
            return { success: false, error: error || 'Failed to promote Stack' };
        }
        log(`Stack "${stack.name}" ${niceId(stack.id)} promoted successfully`);
        const { blueprintFilePath } = blueprint.fileInfo;
        try {
            patchConfigFile(blueprintFilePath, {
                stackId: stack.id,
                organizationId: stack.scopeId,
                projectId: null,
            });
            log('Blueprint configuration updated');
        }
        catch {
            return {
                success: false,
                error: 'Stack promoted successfully but failed to update local Blueprint configuration. No config file found.',
            };
        }
        return { success: true, data: { stack } };
    }
    catch (error) {
        const errorMessage = error instanceof Error ? error.message : String(error);
        return { success: false, error: errorMessage };
    }
}
