// @sanity/runtime-cli/dist/commands/blueprints/promote.js
import { Flags } from '@oclif/core';
import { DeployedStackCommand } from '../../baseCommands.js';
import { blueprintPromoteCore } from '../../cores/blueprints/promote.js';
import { Logger } from '../../utils/logger.js';
export default class PromoteCommand extends DeployedStackCommand {
    static summary = 'Promote a Stack deployment to a broader scope';
    static description = `EXPERIMENTAL! Promotes a Stack, changing its scope from project to organization. This is a one-way trip.`;
    static hidden = true;
    static examples = [
        '<%= config.bin %> <%= command.id %>',
        '<%= config.bin %> <%= command.id %> --stack <name-or-id>',
    ];
    static flags = {
        stack: Flags.string({
            description: 'Stack name or ID to promote',
            aliases: ['id'],
        }),
        'i-know-what-im-doing': Flags.boolean({
            description: 'Must be set',
            default: false,
        }),
        force: Flags.boolean({
            description: 'Skip confirmation prompt',
            default: false,
        }),
    };
    async run() {
        const { success, error, data } = await blueprintPromoteCore({
            bin: this.config.bin,
            log: Logger(this.log.bind(this), this.flags),
            token: this.sanityToken,
            blueprint: this.blueprint,
            stackId: this.stackId,
            scopeType: this.scopeType,
            scopeId: this.scopeId,
            deployedStack: this.deployedStack,
            auth: this.auth,
            validateResources: this.flags['validate-resources'],
            flags: this.flags,
        });
        if (!success)
            this.error(error);
        return data;
    }
}
